### PR TITLE
Switch to buildx to build airflow images

### LINKED ISSUE #20258: Switch to BUILDX to build images
We should use buildx to build and cache our image

https://docs.docker.com/buildx/working-with-buildx/

In the past it was difficult because we are heavily use caching of the imges and the cache of BUILDX was not compatible with docker registry. Howvever with moving to GHCR.io and improvements in Buildx infrastructure it should now be possible to push and pull cache which will be both BUILDX and regular docker build compatible.

### PR DESCRIPTION
**Pull request by potiuk**
The "buildkit" is much more modern docker build mechanism and supports
multiarchitecture builds which makes it suitable for our future ARM
support, it also has nicer UI and much more sophisticated caching
mechanisms as well as supports better multi-segment builds.

BuildKit has been promoted to official for quite a while and it is
rather stable now. Also we can now install BuildKit Plugin to docker
that add capabilities of building and managin cache using dedicated
builders (previously BuildKit cache was managed using rather
complex external tools).

This gives us an opportunity to vastly
simplify our build scripts, because it has now much more robust caching
mechanism than the old docker build (which forced us to pull images
before using them as cache).

We had a lot of complexity involved in efficient caching
but with BuildKit all that can be vastly simplified and we can
get rid of:

  * keeping base python images in our registry
  * keeping build segments for prod image in our registry
  * keeping manifest images in our registry
  * deciding when to pull or pull&build image (not needed now, we can
    always build image with --cache-from and buildkit will pull cached
    layers as needed
  * building the image when performing pre-commit (rather than that
    we simply encourage users to rebuild the image via breeze command)
  * pulling the images before building
  * separate 'build' cache kept in our registry (not needed any more
    as buildkit allows to keep cache for all segments of multi-segmented
    build in a single cache
  * the nice animated tty UI of buildkit eliminates the need of manual
    spinner
  * and a number of other complexities.

Please take a look only at the last commit as it is based on #20238 #20679

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
This is a VAST simplification of our docker build caching (using modern BuildKit plugin). I have to do some more testing with it, but on top of removing several hundreds of lines of Bash that I implemented when BuildKit did not have all the capabilities needed, it's also very stable and robust and opens the path to multi-architecture images (for ARM). 

I would do some more testing - but I would also love to merge #20238 (optimization of Dockerfiles) that this build is based on.

**Comment by potiuk** [comment]
cc: @Bowrna @edithturn - > that will make your job much easier, it simplifies a lot of the "caching" complexity - that's why I did not want you to start looking at it before as I knew we are going to get it WAAAAY simpler with BuildKit.

**Comment by uranusjr** [comment]
I’m not too familiar with buildx in the first palce and will dismiss my request for review.

**Comment by ashb** [comment]
Whats buildx vs buildkit?

**Comment by potiuk** [comment]
> Whats buildx vs buildkit?

Essentially the same. Buildx is a plugin you can install in docker to get more capabilities of buildkit available as "docker buildx build" command. You do not need it run "buildkit" builds (`DOCKER_BUILDKIT=1` is enough in docker 1.18+) but you need it for example to build and push cache to github registry.

An example of that is our prod image. It is multi-segmented image so in order to prepare a good cache for the builds I need to do it with command smilar to `docker buildx build . --cache-to=type=registry,ref=ghcr.io/,........,mode=max`

This was (about a year ago when I checked last time) missing in buildkit - you had to do some strange combination of not-yet-released-then tools from "moby" to build and refresh the cache- but now with the plugin it's a "breeze" to manage and prepare the cache.

In our case users will not have to install buildx plugin, but it will have to be available (eventually) on our self-hosted runners to refresh the cache on main builds (I will add it) and any time you want to manually refresh the cache with `./breeze prepare-build-cache`. 

https://docs.docker.com/engine/reference/commandline/buildx_build/

**Comment by potiuk** [comment]
It also has all the nice things about building multi-platform images. The nice thing about buldx is that you can have mutliple builders - for example different builders for different platforms, or you could have " build cache server" where you builds such caches and organize it in the way that each builder is completely separated from the "docker engine" it runs on. It works in the way that each builder runs as a separate container and has private image 'storage" so when the builder builds an image, it's not visible via `docker images`. 

Pretty nice solution to organize your builds when you have mutli-platform, multi-branch, multil-whatever case. Initially seems much more complex than original docker build system, but it is actually very intuitive.

**Comment by potiuk** [comment]
BTW. This is how image building will look like when this one is merged:

https://user-images.githubusercontent.com/595491/148693440-2af4dc38-06e2-4fd5-8bab-4d21170ddea2.mp4

It's a great improvement - both UI and speed/performance wise. No need to pull the whole image before using it as a cache, it will automatically determine which layers can be pulled an which needs to be rebuilt, it will use multiple processors for rebuilding when more layers are rebuilt, it will do everything in one step ! And it is supported by codespaces natively too, so we will be able to make the image works for codespaces (PR on that is coming too).

**Comment by ashb** [comment]
FYI "Video can't be played because the file is corrupt"

File: Dockerfile
Comment (mik-laj) [inline Dockerfile]: Should we add information about frontend?
```
# syntax = docker/dockerfile:1.3
```

**Comment by potiuk** [comment]
> FYI "Video can't be played because the file is corrupt"

Works for me in Chrome :)

File: Dockerfile
Comment (potiuk) [reply Dockerfile]: > Should we add information about frontend?

Good idea: 

> To use the external frontend, the first line of your Dockerfile needs to be # syntax=docker/dockerfile:1.3 pointing to the specific image you want to use.

> BuildKit also ships with Dockerfile frontend builtin but it is recommended to use an external image to make sure that all users use the same version on the builder and to pick up bugfixes automatically without waiting for a new version of BuildKit or Docker engine.

Comment (potiuk) [reply Dockerfile]: Added

File: Dockerfile.ci
Comment (mik-laj) [inline Dockerfile.ci]: Should we add a frontend definition to this file also?
```
# syntax=docker/dockerfile:1.3
```

**Comment by potiuk** [comment]
Right! This is the last one that remained from the image optimization steps - switching to buildx/buildkit is a major improvement int the way how our images are being built. I am going to test also if our Self-hosted images will cope well with building and pushing the cache (I need to make sure  that buildx plugin is installed for it). 

Really looking forward to this one merged as it will help with building ARM images a lot and will improve the ./breeze experience of all the breeze users a lot (not mentioning -500 lines of bash code that was only needed because before buidkit caching I had to re-implement pretty much all caching experience for "interactive use case" with breeze.

BTW. I am super happy that this code can be removed - it was total pain in the neck to add, maintain and fix it and implement all the workarounds just to overcome the limitation of the original docker caching schemes. Fortunately docker people did a great job in the way how caching has been implemented with buildkit and the buildx plugin allows us to manage it efficiently and automatically.

After this is merged and works for a few days we will be able to remove half of our images we keep in ghcr.io :D

**Comment by potiuk** [comment]
Actually, we will be able to remove 3/5 of our airflow images from ghcr.io:

* Python base images
* Build images
* Manifest images 

The only remaining ones will be plain PROD and CI images.

All those we will be able to remove (after I cherry-pick those changes to 2.2 line)

File: Dockerfile
Comment (potiuk) [inline Dockerfile:331]: BTW. We do not need those labels any more because we are not pushing the 'build' segment to our repo any more. This yet another great feature of buildkit cache, that you can built it with "mode=max" where the cache will contain layers for all segments - the final and intermediate ones.

File: IMAGES.rst
Comment (potiuk) [inline IMAGES.rst:262]: * 3/5 of the images can be scraped now. Just :cache for PROD and CI are enough to efficiently rebuild the image.

File: IMAGES.rst
Comment (potiuk) [inline IMAGES.rst:545]: No need for image manifest either :). We needed it to be able to determine how "big" the difference is of the current Dockerfile /image we had locally vs the one available as cache. This was mainly because of catch-22 - you had to pull the image first to use it as cache, and you could not really inspect the content of the remote image before pulling it without authenticating  (no public API for that). The manifest was really implemented to overcome that limitation.

With Buildkit  cache, each layer is checked separataly and independently from each other without actually pulling the image. This is nicely visualized during the build process where CACHE/REBUILD status is shown very quickly without actually pulling any image. This way we completely do not have to worry how "big" the difference is vs locally rebuilt image. If the difference is "big" the necessary layers will be pulled, if we have modified the image/copied files in the way that they are different than cache - they will be rebuilt. All super-optimised.

File: breeze
Comment (potiuk) [inline breeze:1018]: No need to force-pull the image or to pull base python image. We will not let the user (by default) to rebuild the image before rebasing to latest main (it can be forced but it fails by default) so pulling latest python image and latest cache should always be good.

File: breeze
Comment (potiuk) [inline breeze:3539]: We do not need to push image from breeze any more. 'prepare cache' takes care about it via --cache-to directive during the build.

File: scripts/ci/libraries/_build_images.sh
Comment (potiuk) [inline scripts/ci/libraries/_build_images.sh:116]: Rather than handling all the different cases when user is not rebased, we now simply (by default) check and require the branch of the user to be rebased to latest `main` when rebuilding the image (and warn the user that not doing so might take a long time). This cuts down the number of cases that we had to handle and eliminates the need of storing python base image in ghcr.io (because different branches and  locally built image  could have different base image).

By asking the user to rebase when rebuilding we make sure that the same base python image is used (and also it makes it possible to use `--pull` flag on all rebuilds to always pull the latest base python image. 

This might have a side effect only for a very short while when the new image is relased and no "main" build succeeded yet - but with the recent stability improvements we had in flaky tests, this case should be very limited.

File: scripts/ci/libraries/_build_images.sh
Comment (potiuk) [inline scripts/ci/libraries/_build_images.sh:257]: All this is gone when we have no manifest image.

File: scripts/ci/libraries/_md5sum.sh
Comment (potiuk) [inline scripts/ci/libraries/_md5sum.sh:155]: We do not need to pull the image at all now - buildkit caching is checked without pulling the image first.

File: scripts/ci/libraries/_push_pull_remove_images.sh
Comment (potiuk) [inline scripts/ci/libraries/_push_pull_remove_images.sh:69]: Same here - all that was needed to check and pull the right images before we could use them as cache. This can be removed now completely.

File: scripts/ci/libraries/_spinner.sh
Comment (potiuk) [inline scripts/ci/libraries/_spinner.sh:1]: Rather than showing our animated spinner during the build I chose the option to simply show the buildkit output directly in the terminal. It's nice, colorful, animated and optimized to show only relevant information and take as little space as possible during the build. So we can remove the animated spinner now. 

Also during the pre-commit  I opted to just show the warning that image is not updated rather that try to build it. User can build it when entering breeze or by (suggested) manual action. This simplifies interaction in pre-commit (there the difficulty is that the terminal there is difficult to get hold of).

**Comment by potiuk** [comment]
I also added some comments explaining why certain things were there and why we can remove them now :). I hope it will make it easier to review it as some of the decisions/reasoning was only in my head (luckily a lot of that can go to the trash bin now :D)

File: Dockerfile.ci
Comment (potiuk) [reply Dockerfile.ci]: Why not :). Good point.

**Review by mik-laj — APPROVED** [review]

**Comment by potiuk** [comment]
FYI. I will run a few more tests in my fork - just to make sure everything is fine with caching of build-image workflow results and will merge it when I come back from Slovakia (going for a few days) just to make sure to support any problems with Breeze. So still some time for reviews :)

File: scripts/ci/libraries/_build_images.sh
Comment (mik-laj) [inline scripts/ci/libraries/_build_images.sh:631]: Do you consider to use `gha` cache?
> gha type type exports cache through [the Github Actions Cache service API](https://github.com/tonistiigi/go-actions-cache/blob/master/api.md#authentication).

https://github.com/docker/buildx/blob/master/docs/reference/buildx_build.md#-export-build-cache-to-an-external-cache-destination---cache-to

Comment (potiuk) [reply scripts/ci/libraries/_build_images.sh:631]: Yep. I did consider it, but the size of GHA cache is FAR too small for us (10GB will be esily 3x exceeded by the latest CI + prod images, v2-2, v2-3 branch and we have a lot more caches that are directly used by the GHA (for various virtualenvs  and the like). Our images are huge (for a good reason though). Also PROD image cache is far bigger than that prod image itself because it contains the "build" stage cache which is many times bigger than the PROD image.

> Github Actions cache saves both cache metadata and layers to GitHub's Cache service. This cache currently has a size limit of 10GB that is shared accross different caches in the repo. If you exceed this limit, GitHub will save your cache but will begin evicting caches until the total size is less than 10 GB. Recycling caches too often can result in slower runtimes overall.

**Pull request merged**
